<!doctype html>
<html lang="en">
<head>
    <!-- Required meta tags -->
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">

    <!-- Bootstrap CSS -->
    <link rel="stylesheet" href="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/css/bootstrap.min.css"
          integrity="sha384-ggOyR0iXCbMQv3Xipma34MD+dH/1fQ784/j6cY/iJTQUOhcWr7x9JvoRxT2MZw1T" crossorigin="anonymous">

    <link rel="stylesheet" href="https://use.fontawesome.com/releases/v5.0.13/css/all.css"
          integrity="sha384-DNOHZ68U8hZfKXOrtjWvjxusGo9WQnrNx2sqG0tfsghAvtVlRW3tvkXWZh58N9jp" crossorigin="anonymous">

    <title>Gwas2VCF</title>
</head>
<body>
<!-- Optional JavaScript -->
<!-- jQuery first, then Popper.js, then Bootstrap JS -->
<script src="https://code.jquery.com/jquery-3.4.1.min.js" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.14.7/umd/popper.min.js"
        integrity="sha384-UO2eT0CpHqdSJQ6hJty5KVphtPhzWj9WO1clHTMGa3JDZwrnQq4sF86dIHNDz0W1"
        crossorigin="anonymous"></script>
<script src="https://stackpath.bootstrapcdn.com/bootstrap/4.3.1/js/bootstrap.min.js"
        integrity="sha384-JjSmVgyd0p3pXB1rRibZUAYoIIy6OrQ6VrjIEaFf/nJGzIxFDsf4x0xIM+B07jRM"
        crossorigin="anonymous"></script>
<script async defer src="https://buttons.github.io/buttons.js"></script>
<style>
    div.hidden {
        display: none;
    }
</style>
<nav class="navbar navbar-expand-sm navbar-dark bg-dark" style="margin-bottom: 10px;">
    <a class="navbar-brand" href="https://github.com/mrcieu/gwas2vcf">Gwas2VCF</a>
    <button class="navbar-toggler" type="button" data-toggle="collapse" data-target="#navbarSupportedContent" aria-controls="navbarSupportedContent" aria-expanded="false" aria-label="Toggle navigation">
        <span class="navbar-toggler-icon"></span>
    </button>
    <div class="collapse navbar-collapse" id="navbarSupportedContent">
        <ul class="navbar-nav mr-auto">
            <li class="nav-item">
                <a class="nav-link" href="#requirements-header">Requirements<span class="sr-only">(current)</span></a>
            </li>
            <li class="nav-item">
                <a class="nav-link" href="#usage-header">Usage</a>
            </li>
            <li class="nav-item">
                <a class="nav-link" href="#upload-header">Upload</a>
            </li>
            <li class="nav-item">
                <a class="nav-link" href="#status-header">Status</a>
            </li>
        </ul>
        <ul class="navbar-nav">
            <li class="nav-item" style="margin-top: 8px;">
                <a class="github-button" href="https://github.com/mrcieu/gwas2vcf" data-icon="octicon-star" data-size="large" data-show-count="true" aria-label="Star mrcieu/gwas2vcf on GitHub">Star</a>
            </li>
            <li class="nav-item" style="margin-top: 8px;">
                <a class="github-button" href="https://github.com/mrcieu/gwas2vcf/issues" data-size="large" data-show-count="true" aria-label="Issue mrcieu/gwas2vcf on GitHub">Issue</a>
            </li>
      </ul>
    </div>
</nav>
<div class="container">
    <h3>Convert GWAS summary statistics to the variant call format</h3>
    <p>A common format to store GWAS summary data and associated metadata to aid with data sharing and tool development (manuscript in preparation)</p>
    <p>The <a target="_blank" href="https://github.com/mrcieu/gwas-vcf-spec">VCF</a> format fulfils the following requirements:</p>
    <ul>
        <li>Pre-existing, well known and well defined format</li>
        <li>Handles complex variants, build differences and multi-allelics using the <a target="_blank" href="https://github.com/samtools/htslib">HTSLIB</a> library</li>
        <li>Indexing makes querying by position extremely fast: <a target="_blank" href="https://github.com/MRCIEU/gwas-vcf-performance">VCF is 8.6-45.5x faster to extract variant(s) by genomic position</a></li>
        <li>Many established tools already exist for manipulation</li>
        <li>The file format is relatively small</li>        
        <li>Indexing time is very fast</li>
        <li>Multiple GWAS may be stored in a single file</li>
        <li>Easy to export data into other tabular formats</li>
        <li>Can import directly into a distributed database: <a target="_blank" href="https://github.com/GenomicsDB/GenomicsDB">GenomicsDB</a></li>
    </ul>
    <h5 id="requirements-header">Requirements</h5>
    <p>This tool expects GWAS summary statistics in tabular format, one variant per row. The minimum fields required are:</p>
    <ul>
        <li>Chromosome</li>
        <li>Base position</li>
        <li>Allele</li>
        <li>Effect size</li>
        <li>Standard error</li>
        <li>P value</li>
    </ul>
    <h5 id="usage-header">Usage</h5>
    <p>Users may run the tool on their own server by downloading <a target="_blank" href="https://github.com/MRCIEU/gwas2vcf">Gwas2VCF</a></p>
    <p>For online submissions:</p>
    <ol>
        <li>Fill in the form below, provide as much information as possible</li>
        <li>Record the unique job identifier</li>
        <li>Check job status</li>
        <li>Download your processed data</li>
    </ol>
    <div class="card" id="submit-form">
        <h5 class="card-header" id="upload-header">Upload</h5>
        <div class="card-body">
            <form class="needs-validation" novalidate>
                Enter file column numbers
                <hr>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="chr_col">Chromosome</label>
                        <input type="number" class="form-control" id="chr_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="pos_col">Base position</label>
                        <input type="number" class="form-control" id="pos_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="ea_col">Effect allele</label>
                        <input type="number" class="form-control" id="ea_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="oa_col">Non-effect allele</label>
                        <input type="number" class="form-control" id="oa_col" required>
                    </div>
                </div>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="beta_col">Effect size</label>
                        <input type="number" class="form-control" id="beta_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="se_col">Standard error</label>
                        <input type="number" class="form-control" id="se_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="pval_col">P-value</label>
                        <input type="number" class="form-control" id="pval_col" required>
                    </div>
                    <div class="col-sm">
                        <label for="ncase_col">Case sample size</label>
                        <input type="number" class="form-control" id="ncase_col">
                    </div>
                </div>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="snp_col">DBSNP rs-identifer</label>
                        <input type="number" class="form-control" id="snp_col">
                    </div>
                    <div class="col-sm">
                        <label for="eaf_col">Effect allele frequency</label>
                        <input type="number" class="form-control" id="eaf_col">
                    </div>
                    <div class="col-sm">
                        <label for="oaf_col">Other allele frequency</label>
                        <input type="number" class="form-control" id="oaf_col">
                    </div>
                    <div class="col-sm">
                        <label for="imp_z_col">Imputation Z score</label>
                        <input type="number" class="form-control" id="imp_z_col">
                    </div>
                </div>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="imp_info_col">Imputation INFO score</label>
                        <input type="number" class="form-control" id="imp_info_col">
                    </div>
                    <div class="col-sm">
                        <label for="ncontrol_col">Control/total sample size</label>
                        <input type="number" class="form-control" id="ncontrol_col">
                    </div>
                </div>
                <hr>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="cohort_cases">Total number of cases used in study</label>
                        <input type="number" class="form-control" id="cohort_cases">
                    </div>
                    <div class="col-sm">
                        <label for="cohort_controls">Total number of controls used in study</label>
                        <input type="number" class="form-control" id="cohort_controls">
                    </div>
                </div>
                <div class="form-row">
                    <div class="col-sm">
                        <label for="header">Genome build</label>
                        <select class="custom-select" id="build">
                            <option value="GRCh37">GRCh37/hg19</option>
                        </select>
                    </div>
                    <div class="col-sm">
                        <label for="delimiter">Column delimiter for file</label>
                        <select class="custom-select" id="delimiter">
                            <option value="tab">Tab-separated</option>
                            <option value="comma">Comma-separated</option>
                            <option value="space">Space-separated</option>
                        </select>
                    </div>
                    <div class="col-sm">
                        <label for="header">Does the file have a header line?</label>
                        <select class="custom-select" id="header">
                            <option value="True">True</option>
                            <option value="False">False</option>
                        </select>
                    </div>
                    <div class="col-sm">
                        <label for="gzipped">Is the file compressed with gzip?</label>
                        <select class="custom-select" id="gzipped">
                            <option value="False">False</option>
                            <option value="True">True</option>
                        </select>
                    </div>
                </div>
                <hr>
                <div class="custom-file">
                    <input type="file" class="custom-file-input" id="gwas_file">
                    <label class="custom-file-label" for="gwas_file">Choose file</label>
                </div>
                <div class="form-group">
                    <div class="form-check">
                        <input class="form-check-input" type="checkbox" value="" id="invalidCheck" required>
                        <label class="form-check-label" for="invalidCheck">
                            I agree the host will not be responsible for loss or corruption of data
                        </label>
                        <div class="invalid-feedback">
                            You must agree before submitting.
                        </div>
                    </div>
                </div>
                <a id="submit-job-btn" type="button" class="btn btn-secondary" style="color:white;">Submit form</a>
                <div id="upload-spinner" class="spinner-border text-primary hidden" role="status">
                    <span class="sr-only">Loading...</span>
                  </div>
            </form>
        </div>
    </div>
    <p></p>
    <div class="card">
        <h5 class="card-header" id="status-header">Status</h5>
        <div class="card-body">
            <form class="form-inline my-2 my-lg-0">
                <input id="search-job-val" class="form-control mr-sm-2" type="search" placeholder="Job id"
                       aria-label="Search">
                <a id="search-job-btn" type="button" style="color:white;" class="btn btn-secondary">&nbsp;Check</a>
            </form>
            <div id="search-result" class="alert hidden" role="alert"></div>
        </div>
    </div>
    <div id="server-message" class="alert hidden" role="alert"></div>
    <footer class="container">
        <p>Need help? Want to host your own pipeline? Visit <a target="_blank"
                                                               href="https://github.com/MRCIEU/gwas2vcfweb">GitHub</a>
        </p>
        <p>&copy; 2019 University of Bristol. All rights reserved.</p>
    </footer>
</div>
<script>
    $(document).ready(function () {
        $("#search-job-btn").click(function () {
            var job_id = $("#search-job-val").val();

            $.ajax({
                type: 'GET',
                url: "/vcf/status",
                data: {"job_id": job_id},
                success: function (data) {
                    console.log(data);
                    $( "#search-result" ).text( data.message ).removeClass( "alert-danger" ).addClass( "alert-success" ).removeClass( "hidden" );
                    if (data['vcf_uri']){
                        $( "#search-result" ).append( " Click to download <a href=\"" + data.vcf_uri + "\">VCF</a> and <a href=\"" + data.tbi_uri + "\">Index</a>");
                    }
                },
                error: function (data) {
                    console.log(data);
                    $( "#search-result" ).text( data.responseText ).removeClass( "alert-success" ).addClass( "alert-danger" ).removeClass( "hidden" );
                }
            });

        });

        $("#submit-job-btn").click(function () {

            // get variables
            var chr_col = $("#chr_col").val();
            var pos_col = $("#pos_col").val();
            var ea_col = $("#ea_col").val();
            var oa_col = $("#oa_col").val();
            var beta_col = $("#beta_col").val();
            var se_col = $("#se_col").val();
            var pval_col = $("#pval_col").val();
            var delimiter = $("#delimiter").val();
            var header = $("#header").val();
            var build = $("#build").val();
            var gzipped = $("#gzipped").val();

            // optional
            var ncase_col = $("#ncase_col").val();
            var snp_col = $("#snp_col").val();
            var eaf_col = $("#eaf_col").val();
            var oaf_col = $("#oaf_col").val();
            var imp_z_col = $("#imp_z_col").val();
            var imp_info_col = $("#imp_info_col").val();
            var ncontrol_col = $("#ncontrol_col").val();
            var cohort_cases = $("#cohort_cases").val();
            var cohort_controls = $("#cohort_controls").val();

            // check numerics
            if (!$.isNumeric(chr_col)) {
                alert("You must provide a numeric column number for chromosome");
                return;
            }
            if (!$.isNumeric(pos_col)) {
                alert("You must provide a numeric column number for position");
                return;
            }
            if (!$.isNumeric(ea_col)) {
                alert("You must provide a numeric column number for effect allele");
                return;
            }
            if (!$.isNumeric(oa_col)) {
                alert("You must provide a numeric column number for other allele");
                return;
            }
            if (!$.isNumeric(beta_col)) {
                alert("You must provide a numeric column number for effect size");
                return;
            }
            if (!$.isNumeric(se_col)) {
                alert("You must provide a numeric column number for SE");
                return;
            }
            if (!$.isNumeric(pval_col)) {
                alert("You must provide a numeric column number for P value");
                return;
            }
            if (!$("#invalidCheck").is(':checked')) {
                alert("You must agree to our terms");
                return;
            }

            formdata = new FormData();
            formdata.append("chr_col", chr_col);
            formdata.append("pos_col", pos_col);
            formdata.append("ea_col", ea_col);
            formdata.append("oa_col", oa_col);
            formdata.append("beta_col", beta_col);
            formdata.append("se_col", se_col);
            formdata.append("pval_col", pval_col);
            formdata.append("delimiter", delimiter);
            formdata.append("header", header);
            formdata.append("build", build);
            formdata.append("gzipped", gzipped);

            // get file
            file = $('#gwas_file').prop('files')[0];
            if (file === undefined){
                alert("You must provide a file");
            } else {
                formdata.append("gwas_file", file);
            }

            if ($.isNumeric(ncase_col)) {
                formdata.append("ncase_col", ncase_col);
            }
            if ($.isNumeric(snp_col)) {
                formdata.append("snp_col", snp_col);
            }
            if ($.isNumeric(eaf_col)) {
                formdata.append("eaf_col", eaf_col);
            }
            if ($.isNumeric(oaf_col)) {
                formdata.append("oaf_col", oaf_col);
            }
            if ($.isNumeric(imp_z_col)) {
                formdata.append("imp_z_col", imp_z_col);
            }
            if ($.isNumeric(imp_info_col)) {
                formdata.append("imp_info_col", imp_info_col);
            }
            if ($.isNumeric(ncontrol_col)) {
                formdata.append("ncontrol_col", ncontrol_col);
            }
            if ($.isNumeric(cohort_cases)) {
                formdata.append("cohort_cases", cohort_cases);
            }
            if ($.isNumeric(cohort_controls)) {
                formdata.append("cohort_controls", cohort_controls);
            }

            $.ajax({
                type: 'POST',
                url: "/txt/upload",
                data: formdata,
                processData: false,
                contentType: false,
                success: function (data) {
                    console.log(data);
                    $( "#upload-spinner" ).addClass( "hidden" );
                    $("#submit-job-btn").prop('disabled', false);
                    $( "#server-message" ).text( "Submitted job " + data.job ).addClass( "alert-success" ).removeClass( "hidden" );
                },
                error: function (data) {
                    console.log(data);
                    $( "#upload-spinner" ).addClass( "hidden" );
                    $("#submit-job-btn").prop('disabled', false);
                    $( "#server-message" ).text( data.responseText ).addClass( "alert-danger" ).removeClass( "hidden" );
                },
                beforeSend: function() {
                    $("#submit-job-btn").prop('disabled', true);
                    $( "#upload-spinner" ).removeClass( "hidden" );
                }
            });

        });
    });






</script>
<script>
// Example starter JavaScript for disabling form submissions if there are invalid fields
(function() {
  'use strict';
  window.addEventListener('load', function() {
    // Fetch all the forms we want to apply custom Bootstrap validation styles to
    var forms = document.getElementsByClassName('needs-validation');
    // Loop over them and prevent submission
    var validation = Array.prototype.filter.call(forms, function(form) {
      form.addEventListener('submit', function(event) {
        if (form.checkValidity() === false) {
          event.preventDefault();
          event.stopPropagation();
        }
        form.classList.add('was-validated');
      }, false);
    });
  }, false);
})();






</script>
</body>
</html>
